feat(AlgebraicTopology): delaborators for truncated simplicial notations

gio256 opened the pull request:
We add delaborators for the following notations, introduced in #20688:

- `⦋m⦌ₙ`, which denotes the `m`-dimensional simplex in the `n`-truncated simplex category.
- `X _⦋m⦌ₙ`, which denotes the `m`-th term of an `n`-truncated simplicial object `X`.
- `X ^⦋m⦌ₙ`, which denotes the `m`-th term of an `n`-truncated cosimplicial object `X`.

If `pp.proofs` is set to `true`, we also pretty-print the proof `p : m ≤ n` for all three notations as `⦋m, p⦌ₙ`, `X _⦋m, p⦌ₙ`, and `X ^⦋m, p⦌ₙ`, respectively.

Credit to @kmill for one piece of code and much metaprogramming inspiration.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->
- [x] depends on: #20688 
- [x] depends on: #23018

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

gio256 committed:
Add truncated simplicial notations

gio256 committed:
Update scripts/noshake.json

gio256 committed:
Formatting

gio256 committed:
Formatting: indentation

gio256 committed:
Move failure message into trunc tactic

gio256 committed:
Add delaborator for SimplexCategory notation [m]ₙ

gio256 committed:
Tweak SimplexCategory.delabMkNotation

gio256 committed:
Add delaborators for SimplicialObject notations

gio256 committed:
doc: docstrings for SimplicialObject delaborators

gio256 committed:
Clarify behavior of notations wrt metavariables

gio256 commented:
The following test, taken from [`#17732`](https://github.com/leanprover-community/mathlib4/pull/17732/files#diff-40ab7cd5656f9da110091d345b5b90bc73d7183951a1a663c173a51a585acca2R24-R31), shows that our delaborator attempts to pretty-print some terms even when the truncation level cannot be subscripted. 

I'm not aware of a fix for this that would still allow pretty-printing things like `[m]ₙ₊₁`, but any suggestions would be appreciated. It also might not matter, as I can't think of any cases where the difference between  `[m]ₙ` and `[m]n✝` would be all that confusing.

```lean
import Mathlib.AlgebraicTopology.SimplexCategory
open SimplexCategory.Truncated

variable (n : ℕ) (m : ℕ) (h : m ≤ n) (x) (hx : x = [m]ₙ) (n : True)

/- The delaborator should ideally not fire here because the truncation level is now `n✝`,
which cannot be subscripted. -/
/-- info: hx : x = [m]n✝ -/
#guard_msgs in #check hx
```

gio256 committed:
fix(SimplexCategory): check valid subscript chars

gio256 committed:
fix(SimplicialObject): check valid subscript chars

gio256 committed:
chore: Rename Meta.subscript and reduce footprint

gio256 committed:
chore: style fixes

gio256 committed:
doc: backport changes from branch gio/trunc-delab

gio256 committed:
fix: scoped macro_rules

gio256 committed:
Merge branch 'master' into gio/trunc-macro

gio256 committed:
doc: fix typo

gio256 committed:
Merge branch 'master' into gio/trunc-delab

gio256 committed:
chore: style fixes

gio256 committed:
chore: formatting of macro_rules

gio256 committed:
fix: backport changes from branch gio/trunc-delab

gio256 committed:
fix: cheaper trunc tactic (defer to omega sooner)

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
chore: Update scripts/noshake.json

gio256 committed:
fix: improve failure message for trunc tactic

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
Merge branch 'master' into gio/trunc-macro

gio256 committed:
style: fix trailing whitespace

gio256 committed:
feat: notation ⦋m⦌ₙ for truncated simplex category

gio256 committed:
chore: rm (priority := high) from notations

gio256 committed:
feat: X ^[m]ₙ for truncated cosimplicial objects

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
test: fix truncated simplicial notation tests

gio256 committed:
Merge branch 'master' into gio/trunc-macro

gio256 committed:
feat: notations X _⦋m⦌ₙ and X ^⦋m⦌ₙ

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
fix: update delaborators for new notations

gio256 committed:
Merge branch 'master' into gio/trunc-macro (partial)

gio256 committed:
fix: merge conflict in SimplexCategory/Defs.lean

gio256 committed:
fix: update scripts/noshake.json

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
fix: update scripts/noshake.json

gio256 committed:
fix: update MathlibTest/SimplexCategory.lean

kim-em commented on Mathlib/AlgebraicTopology/SimplexCategory/Defs.lean:
```diff
@@ -167,6 +176,74 @@ noncomputable def inclusion.fullyFaithful (n : ℕ) :
+section Meta
+
+/-- Some quick attempts to prove that `⦋m⦌` is `n`-truncated (`⦋m⦌.len ≤ n`). -/
+macro "trunc" : tactic =>
+  `(tactic| first | decide | assumption |
+    dsimp only [SimplexCategory.len_mk]; omega)
```
I remember this being discussed elsewhere, but can you remind me why `decide` and `assumption` are needed here?

gio256 replied:
They are not needed (in that they don't solve any goals that aren't solved by `omega`), but I decided to keep them after some rudimentary benchmarking. Most of the goals solved by `trunc` can be solved by `decide` or `assumption`, and it is also nicer to carry around a single hypothesis rather than an omega proof.

That being said, I am happy to replace this with a simple `dsimp only [SimplexCategory.len_mk]; omega`.

kim-em replied:
In fact, can you use `get_elem_tactic`? If that works well for your applications, that is probably the most robust with respect to future changes in tactics. But otherwise just `omega` is best, I think.

gio256 replied:
`get_elem_tactic` was a perfect replacement for `trunc`, thanks!

gio256 replied:
Actually, `get_elem_tactic` turned out to be slightly less effective than `omega` in some cases. I have restored the `trunc` tactic, but it just delegates to `omega` now.

kim-em commented on Mathlib/AlgebraicTopology/SimplexCategory/Defs.lean:
```diff
@@ -167,6 +176,74 @@ noncomputable def inclusion.fullyFaithful (n : ℕ) :
 theorem Hom.ext {n} {a b : Truncated n} (f g : a ⟶ b) :
     f.toOrderHom = g.toOrderHom → f = g := SimplexCategory.Hom.ext _ _
 
+section Meta
```
Could I encourage moving this to a separate file, perhaps even `Mathlib/Tactics/AlgebraicTopology/SimplexCategory.lean`?

kim-em commented on Mathlib/AlgebraicTopology/SimplicialObject/Basic.lean:
```diff
@@ -245,6 +252,47 @@ def whiskering {n} (D : Type*) [Category D] : (C ⥤ D) ⥤ Truncated C n ⥤ Tr
 
+section Meta
+
+open Mathlib.Tactic (subscriptTerm) in
```
Please add cross-references to the other macro/delaborator, and make sure both clearly describe their scope.

kim-em commented on MathlibTest/SimplexCategory.lean:
```diff
@@ -0,0 +1,76 @@
+
+variable {x} (hx : x = ⦋m⦌ₙ) (n : True)
+
+/- The delaborator should fail because `n` is now shadowed and `✝` cannot be
+subscripted. -/
```
This feels like a reason to not do this, rather than to have tests that guard the broken behaviour!

gio256 replied:
Similar to my comment above, I think this test exists to make sure that the delaborator doesn't create any confusion in extreme cases, but I don't actually expect these cases to arise very often in practice.

If we expect to be regularly using arbitrary expressions for the truncation level, then we should definitely change the notation.

gio256 commented on MathlibTest/SimplexCategory.lean:
```diff
@@ -0,0 +1,76 @@
+/-- info: ⦋m⦌ₙ ₊ ₁ : CategoryTheory.FullSubcategory fun a => a.len ≤ n + 1 -/
+#guard_msgs in #check ⦋m⦌ₙ₊₁
+
+/-- info: ⦋m⦌ₙ ₊ ₍₄ ₋ ₍₂ ₊ ₁₎₎ :
```
To be clear, I have no particular attachment to this notation. I'm happy to change it if that's what we want to do.

However, I don't think this test is indicative of how the notation will actually be used. I expect that the vast majority of the time the truncation level (which is subscripted) will be a numeric literal or an extremely simple expression, perhaps as complex as `n + 1`.

gio256 replied:
@kim-em I believe I have addressed your other concerns above, but please let me know if you would prefer that we consider alternatives to the notations defined here. I am happy to start such a discussion on zulip.

gio256 committed:
fix: make trunc tactic scoped

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
fix: replace trunc tactic with get_elem_tactic

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
fix: move truncated simplex category notations

gio256 committed:
style: reorganize imports

gio256 committed:
doc: cross-ref (co)simplicial object notations

gio256 committed:
fix: rm unused import

gio256 committed:
style: clean up diff

gio256 committed:
fix: restore (simplified) trunc tactic

gio256 committed:
chore: update scripts/noshake.json

gio256 committed:
chore: update scripts/noshake.json

Note that this addition is not strictly necessary. We could
alternatively omit the import of `Mathlib.Tactic.SimplexCategory` in
`Mathlib.AlgebraicTopology.SimplexCategory.Basic` and explicitly import
`Mathlib.Tactic.SimplexCategory` when we wish to use the truncated
simplicial notations. However, this commit instead allows all importers
of `Mathlib.AlgebraicTopology.SimplexCategory.Basic` to use the
truncated simplicial notations without further imports.

gio256 committed:
Merge branch 'master' into gio/trunc-macro

gio256 committed:
Merge branch 'gio/trunc-macro' into gio/trunc-delab

gio256 committed:
Merge branch 'master' into gio/trunc-delab

gio256 committed:
fix: remove unused import

eric-wieser commented on Mathlib/Tactic/SimplexCategory.lean:
```diff
@@ -0,0 +1,85 @@
+open Mathlib.Tactic.Superscript (Mapping)
+open Lean PrettyPrinter.Delaborator SubExpr
+
+/-- Checks that the provided expression can be subscripted. -/
+private partial def subscriptable (e : Expr) : DelabM Unit := do
+  -- Any number or free variable with a subscriptable name is subscriptable.
+  if (← name e).any isSubscriptable || (← delab) matches `($_:num) then return
+  -- Addition and subtraction are subscriptable if their operands are.
+  guard <| e.isAppOfArity ``HAdd.hAdd 6 || e.isAppOfArity ``HSub.hSub 6
+  let #[_, _, _, _, x, y] := e.getAppArgs | failure
+  let _ ← withNaryArg 4 <| subscriptable x
+  let _ ← withAppArg <| subscriptable y
+where
+  -- Return the user-facing name of any constant or free variable.
+  name : Expr → MetaM (Option Name)
+    | Expr.const name _ => pure name
+    | Expr.fvar name => name.getUserName
+    | _ => pure none
+  -- Return `true` if every character in `s` can be subscripted.
+  isSubscriptable (s : Name) : Bool :=
+    s.toString.toList.all Mapping.subscript.toSpecial.contains
+
+/-- Checks that the provided expression can be subscripted before delaborating. -/
+def Meta.subscript (e : Expr) : Delab := subscriptable e >>= fun () ↦ delab
```
This belongs in the same file as the subscript notation (or at least, next to it)

gio256 replied:
Currently, this `SimplexCategory.Truncated.Meta.subscript` function is used in three different notations: `⦋m⦌ₙ` (defined in this file) as well as `X _⦋m⦌ₙ` and `X ^⦋m⦌ₙ` (defined in `SimplicialObject/Basic.lean`).

Would you suggest moving it to somewhere more generic, or is it fine to be defined in the same place as the first of these three notations? I can also add a reference to this function from the two delaborators in `SimplicialObject/Basic.lean`.

eric-wieser replied:
I mean it should live next to the thing that implements the low level parser for subscripts (`subscriptTerm`?), much further upstream.

gio256 replied:
Oh, I misunderstood. My mistake. 

My only concern is that `subscriptable` may not be general enough to live in `Mathlib.Util`. In particular, it doesn't yet handle function application, which may be valid in a subscript. I will move it and add an explanatory comment, though.

Would it be worth adding an analogous `superscriptable` function as well (perhaps in a separate PR)?

eric-wieser replied:
I think all these new delaborator tools like `subscriptable` belong in a separate PR to this one, and at that point `superscrptable` should come along for the ride.

> In particular, it doesn't yet handle function application, which may be valid in a subscript.

I think it's fine as long as the function is overly conservative, and the docstring mentions where.

gio256 replied:
There are some rough edges still, but #23018 provides a slightly more robust implementation of `subscriptable` and `superscriptable`.

eric-wieser commented on MathlibTest/SimplexCategory.lean:
```diff
@@ -0,0 +1,73 @@
+/-- info: ⦋m⦌ₙ : CategoryTheory.FullSubcategory fun a => a.len ≤ n -/
+#guard_msgs in #check ⦋m⦌ₙ
+
+/-- info: ⦋m⦌ₙ ₊ ₁ : CategoryTheory.FullSubcategory fun a => a.len ≤ n + 1 -/
+#guard_msgs in #check ⦋m⦌ₙ₊₁
```
The test should match the input here:
```suggestion
/-- info: ⦋m⦌ₙ ₊ ₁ : CategoryTheory.FullSubcategory fun a => a.len ≤ n + 1 -/
#guard_msgs in #check ⦋m⦌ₙ ₊ ₁
```
If the spaces aren't in fact legal in the notation, then the delaborator should not emit them.

gio256 committed:
style: spacing in MathlibTest/SimplexCategory.lean

gio256 committed:
feat: delabSubscript and delabSuperscript

gio256 committed:
fix: private function application

gio256 committed:
test: restructure sub/superscript delab tests

gio256 committed:
fix: subscript partially-applied functions

gio256 committed:
test: better syntax for subscript test macro

gio256 committed:
fix: prefer getExpr to passing around expression

gio256 committed:
fix: () : Unit is valid in subscripts

gio256 committed:
fix: camelCase defs

gio256 committed:
test: test partially-applied fn in subscript

gio256 committed:
chore: make sub/superscriptable fns private

gio256 committed:
doc: clean up and comment

gio256 committed:
chore: clarify name function

gio256 committed:
doc: clarify comment

gio256 committed:
chore: imports, comments, add delab section

gio256 committed:
test: add tests that don't trigger the delaborator

gio256 committed:
fix: delab first, then check subscript validity

gio256 committed:
chore: clean up isSuperscriptable

gio256 committed:
chore: combine superscriptable and subscriptable

gio256 committed:
test: rm smul test

gio256 committed:
doc: add comment for Superscript.isValid

gio256 committed:
test: clean up test fn naming

gio256 committed:
test: add nested superscript(subscript) tests

gio256 committed:
doc: fix typo

gio256 committed:
test: add dot notation tests

gio256 committed:
chore: rm unused universe

gio256 committed:
style: use anon dot notation in delabSubscript fns

gio256 committed:
fix: reject values that are already subscripted

gio256 committed:
Merge branch 'master' into gio/subscript

gio256 committed:
Merge branch 'gio/subscript' into gio/trunc-delab

gio256 committed:
chore: rebase on top of #23018

gio256 committed:
chore: clean up Mathlib.Tactic.SimplexCategory

gio256 committed:
chore: clarify Superscript.isValid logic

gio256 committed:
Merge branch 'master' into gio/trunc-delab

gio256 committed:
fix: simplicial delaborator tests

gio256 committed:
style: golf Superscript.isValid valid helper fn

Co-authored-by: Jon Eugster <eugster.jon@gmail.com>

gio256 committed:
style: golf Superscript.isValid scripted helper fn

Co-authored-by: Jon Eugster <eugster.jon@gmail.com>

gio256 committed:
Merge branch 'gio/subscript' into gio/trunc-delab

kmill committed:
restore using decl_name to guarantee SyntaxNodeKind uniqueness

kmill committed:
minimize diff

gio256 committed:
Merge branch 'gio/subscript' into gio/trunc-delab

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

joneugster reviewed (approved):
Hi @gio256! Sorry this PR was lying around for quite a while!

I think this is useful for subscripting `⦋m⦌ₙ` and `⦋m⦌ₙ₊₁` which seem to be the main intended use cases.

As far as I can see the main blocker of the PR was that @kim-em commented, that the tests are covering cases outside of the intended usage. I think this is fine as is.

Could you please resolve merge conflicts, so we can get this merged afterwards?
